src/node.h

```diff
@@ -521,12 +517,120 @@ typedef void (*promise_hook_func) (v8::PromiseHookType type,
+
+/* If the native API doesn't inherit from the helper class then the callbacks
+ * must be triggered manually. This triggers the init() callback. The return
+ * value is the uid assigned to the resource. */
+async_uid EmitAsyncInit(v8::Isolate* isolate,
+                        v8::Local<v8::Object> resource,
+                        const char* name,
+                        async_uid trigger_id = -1);
+
+/* Emit the destroy() callback. */
+void EmitAsyncDestroy(v8::Isolate* isolate, async_uid id);
+
+/* An API specific to emit before/after callbacks is unnecessary because
+ * MakeCallback will automatically call them for you.
+ *
+ * These methods may create handles on their own, so run them inside a
+ * HandleScope. */
+v8::MaybeLocal<v8::Value> MakeCallback(v8::Isolate* isolate,
+                                       v8::Local<v8::Object> recv,
+                                       v8::Local<v8::Function> callback,
+                                       int argc,
+                                       v8::Local<v8::Value>* argv,
+                                       async_uid asyncId,
+                                       async_uid triggerId = -1);
+v8::MaybeLocal<v8::Value> MakeCallback(v8::Isolate* isolate,
+                                       v8::Local<v8::Object> recv,
+                                       const char* method,
+                                       int argc,
+                                       v8::Local<v8::Value>* argv,
+                                       async_uid asyncId,
+                                       async_uid triggerId = -1);
+v8::MaybeLocal<v8::Value> MakeCallback(v8::Isolate* isolate,
+                                       v8::Local<v8::Object> recv,
+                                       v8::Local<v8::String> symbol,
+                                       int argc,
+                                       v8::Local<v8::Value>* argv,
+                                       async_uid asyncId,
+                                       async_uid triggerId = -1);
+
+/* Helper class users can optionally inherit from. If
+ * `AsyncResource::MakeCallback()` is used, then all four callbacks will be
+ * called automatically. */
+class AsyncResource {
+  public:
+    AsyncResource(v8::Isolate* isolate,
+                  v8::Local<v8::Object> resource,
+                  const char* name,
+                  async_uid trigger_id = -1)
+        : isolate_(isolate), resource_(isolate, resource) {
+      uid_ = EmitAsyncInit(isolate, resource, name, trigger_id);
+    }
+
+    ~AsyncResource() {
+      EmitAsyncDestroy(isolate_, uid_);
+    }
+
+    v8::MaybeLocal<v8::Value> MakeCallback(
+        v8::Local<v8::Function> callback,
+        int argc,
+        v8::Local<v8::Value>* argv,
+        async_uid trigger_id = -1) {
+      return node::MakeCallback(isolate_, get_resource(),
+                                callback, argc, argv,
+                                uid_, trigger_id);
```

Comment:
The `trigger_id` passed to `AsyncResource::MakeCallback` should be the same as the one passed to `AsyncResource::AsyncResource`, so I think we should just store it in a private field in `AsyncResource::AsyncResource` and use that field in `AsyncResource::MakeCallback`. That way we don't need the `trigger_id` argument in `AsyncResource::MakeCallback`.